{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 1: Write a Python script that proves that the lines of data in Germplasm.tsv, \n",
    "# and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). \n",
    "# (hint: This will help you decide how to load the data into the database)\n",
    "germplasmfile = open('Germplasm.tsv', 'r')\n",
    "locusgenefile = open('LocusGene.tsv', 'r')\n",
    "\n",
    "germplasmfile.seek(0)\n",
    "locusgenefile.seek(0)\n",
    "\n",
    "for line in germplasmfile.readlines():\n",
    "    line = line.rstrip()\n",
    "    locusg, germplasm, phenotype, pubmed = line.split('\\t')\n",
    "    \n",
    "for line in locusgenefile.readlines() :\n",
    "    line = line.rstrip()\n",
    "    locusl, gene, protein_lenght = line.split('\\t')\n",
    "\n",
    "germplasmfile.seek(0)\n",
    "locusgenefile.seek(0)\n",
    "\n",
    "if locusg == locusl :\n",
    "    print(\"same locusAGIcode\", locusg, locusl)\n",
    "else :\n",
    "    print(\"Not the same!!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1: They present the same locus AGI code in the same sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 2: Design and create de database\n",
    "    # It should have two tables - one for each of the two data files.\n",
    "    # The two tables should be linked in a 1:1 relationship.\n",
    "    # You may use either sqlMagic or pymysql to build the database.\n",
    "    \n",
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creo la base de datos\n",
    "%sql create database exam2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selecciono la base de datos que voy a usar\n",
    "%sql use exam2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creo la tabla germplasm:\n",
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(30) NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(1000) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creo la tabla locus_gene\n",
    "%sql CREATE TABLE locus_gene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(30) NOT NULL, gene VARCHAR(30) NOT NULL, protein_lenght INTEGER NOT NULL, germ_id INTEGER NOT NULL);\n",
    "%sql DESCRIBE locus_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Problem 2 : I have used sqlMagic to create the database, that I have called exam2 and I also have created the two tables with relation 1:1 who will contain the data of the two files. The way I have connected the two tables is adding a germ_id column in the locus_gene table that will contain the germplasm.id of each record."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 3: Using pymysql, create a Python script that reads the data from these files, and fills the database. \n",
    "    # There are a variety of strategies to accomplish this.  \n",
    "    # I will give all strategies equal credit - do whichever one you are most confident with.\n",
    "\n",
    "#FILL TABLE GERMPLASM:\n",
    "    \n",
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2', \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False  \n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv') as csvfile:\n",
    "            germplasm = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "           \n",
    "            for row in germplasm:\n",
    "                print(row)\n",
    "                sql = \"INSERT INTO germplasm(locus, germplasm, phenotype, pubmed) values('{}', '{}', '{}', {})\".format(row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "                cursor.execute(sql)\n",
    "                connection.commit()\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql select * from germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FILL TABLE LOCUS_GENES:\n",
    "\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2', \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False  \n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            locusgene = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            #locusgene = fieldnames (Locus, Gene, ProteinLength)\n",
    "            #germplasm = fieldnames (Locus, germplasm, phenotype, pubmed)\n",
    "            for row in locusgene:\n",
    "                print(row)\n",
    "                locus= row[\"Locus\"]\n",
    "                print(\"SELECT id FROM germplasm WHERE locus ='{}'\".format(locus))\n",
    "                sql = \"SELECT id FROM germplasm WHERE locus ='{}'\".format(locus)\n",
    "                cursor.execute(sql)\n",
    "                results = cursor.fetchall()\n",
    "                germid = results[0]['id']\n",
    "                print(\"Germid is {}\".format(germid))\n",
    "                \n",
    "                sql = \"INSERT INTO locus_gene (locus, gene, protein_lenght, germ_id) values ('{}', '{}', {}, {})\".format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"], germid)\n",
    "                cursor.execute(sql)\n",
    "                connection.commit()\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql select * from locus_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Problem 3: I have filled the two tables using the tsv files that contain the data. For that I have used pymysql and the csv library, because I have opened the two files using the DictReader tool. As the files have headers, I do not use the fieldname option, but I had to write the delimiter as a \\t (and not a ',' because they are separated by tabs). \n",
    "To fill the germplasm table I have used INSERT INTO command from sql and related it to the file.\n",
    "To fill the locus_gene table I have dome the same, but in this case I had to add a column (the germ_id) that I have extracted from the germplasm table. Because of that, I used the command SELECT FROM... WHERE to select the id from gemplasm and then insert it into the locus_gene table using INSERT INTO command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 4: Create the file\n",
    "    \n",
    "import csv\n",
    "import io\n",
    "\n",
    "with open('problem4file.tsv', 'w', newline ='') as csvfile:\n",
    "    problem4filewriter = csv.writer (csvfile, delimiter = \"\\t\", quotechar= '\"')\n",
    "    \n",
    "csvfile.close()\n",
    "csvfile = open(\"problem4file.tsv\")\n",
    "print (csvfile.read())\n",
    "csvfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 4.1 : Create a report that shows the full, joined, \n",
    "               # content of the two database tables (including a header line)\n",
    "\n",
    "%sql SELECT * FROM germplasm, locus_gene WHERE locus_gene.germ_id = germplasm.id \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SENDING THE QUERY TO THE FILE:\n",
    "\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2', \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False  \n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        with open('problem4file.tsv', 'w', newline = \"\") as csvfile:\n",
    "            problem4file = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            sql = \"SELECT * FROM germplasm , locus_gene WHERE locus_gene.germ_id = germplasm.id\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([\"This is the result from the first query of problem 4\"])\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            cursor.execute(sql)\n",
    "            connection.commit()\n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#PROBLEM 4.1: To show everything of the two tables joined together, I have used the comand SELECT * (everything) FROM the two tables and I have added the condition WHERE the column germ_id of the table locus_gene was equal to the id from germplasm table. I can do this because my two tables are conected by their id code. Finally, I have sent to the file 'problem4file' the results from the query.\n",
    "Note: Maybe I could have used the locusID like the primary key when creating the tables but is a larger sequence of characters comparing to the simple number of the id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 4.2: Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "%sql SELECT * FROM locus_gene LEFT JOIN germplasm ON \\\n",
    "    germ_id = germplasm.id WHERE \\\n",
    "    locus_gene.gene = 'SKOR' OR \\\n",
    "    locus_gene.gene = 'MAA3'\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SENDING THE QUERY TO THE FILE:\n",
    "\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2', \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False  \n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        with open('problem4file.tsv', 'a', newline = \"\") as csvfile:\n",
    "            problem4file = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            sql = \"\"\"SELECT * FROM locus_gene LEFT JOIN germplasm ON\n",
    "            germ_id = germplasm.id WHERE locus_gene.gene = 'SKOR' OR locus_gene.gene = 'MAA3'\"\"\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([\"This is the result from the second query of problem 4\"])\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            cursor.execute(sql)\n",
    "            connection.commit()\n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#PROBLEM 4.2: I have used the LEFT JOIN command in sql to show all data from locus_gene table but only the match for the germplasm table where those genes (SKOR and MAA3) are and I connected both tables with the germ_id column from locus_gene table. Finally, I added to the file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 4.3: Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "%sql SELECT COUNT(*) AS \"Number of entries chr1\" FROM germplasm WHERE \\\n",
    "     locus REGEXP 'AT1G.+' \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number of entries chr2\" FROM germplasm WHERE \\\n",
    "     locus REGEXP 'AT2G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number of entries chr3\" FROM germplasm WHERE \\\n",
    "     locus REGEXP 'AT3G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number of entries chr4\" FROM germplasm WHERE \\\n",
    "     locus REGEXP 'AT4G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number of entries chr5\" FROM germplasm WHERE \\\n",
    "     locus REGEXP 'AT5G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SENDING THE QUERY TO THE FILE:\n",
    "\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2', \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False  \n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        with open('problem4file.tsv', 'a', newline = '') as csvfile:\n",
    "            problem4file = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            sql = \"SELECT COUNT(*) AS 'Number of entries chr1' FROM germplasm WHERE locus REGEXP 'AT1G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([\"This is the result from the third question of problem 4\"])\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT COUNT(*) AS 'Number of entries chr2' FROM germplasm WHERE locus REGEXP 'AT2G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT COUNT(*) AS 'Number of entries chr3' FROM germplasm WHERE locus REGEXP 'AT3G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT COUNT(*) AS 'Number of entries chr4' FROM germplasm WHERE locus REGEXP 'AT4G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT COUNT(*) AS 'Number of entries chr5' FROM germplasm WHERE locus REGEXP 'AT5G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            cursor.execute(sql)\n",
    "            connection.commit()\n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#PROBLEM 4.3: I used the command SELECT COUNT() and REGULAR EXPRESSIONS in sql to look for the number of genes related to each chromosome (1-5) and then I added it to the tsv file ussing the append option ('a'). I really know it is an awful code but I didn't find an abstract solution to the problem... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PROBLEM 4.4: Create a report that shows the average protein length \n",
    "            # for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "#For Chromosome 1:\n",
    "%sql SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE \\\n",
    "     locus REGEXP 'AT1G.+'\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For Chromosome 2:\n",
    "%sql SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE \\\n",
    "     locus REGEXP 'AT2G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For Chromosome 3:\n",
    "%sql SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE \\\n",
    "     locus REGEXP 'AT3G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For Chromosome 4:\n",
    "%sql SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE \\\n",
    "     locus REGEXP 'AT4G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For Chromosome 5:\n",
    "%sql SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE \\\n",
    "     locus REGEXP 'AT5G.+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SENDING THE QUERY TO THE FILE:\n",
    "\n",
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2', \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False  \n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "        with open('problem4file.tsv', 'a', newline = '') as csvfile:\n",
    "            problem4file = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            sql = \"SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE locus REGEXP 'AT1G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([\"This is the result from the fourth question of problem 4\"])\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE locus REGEXP 'AT2G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE locus REGEXP 'AT3G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE locus REGEXP 'AT4G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            sql = \"SELECT AVG(locus_gene.protein_lenght) FROM locus_gene WHERE locus REGEXP 'AT5G.+'\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            newdata = results\n",
    "            problem4file.writerow([results])\n",
    "            \n",
    "            cursor.execute(sql)\n",
    "            connection.commit()\n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#PROBLEM 4.4: Almost the same solution as the 4.3 problem, but instead of using the COUNT command this time I have used the AVG command."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
